### PR TITLE
Winograd Update for Small batches 

### PR DESCRIPTION
**Pull request by jagrit06**
## Proposed changes

- Build in padding to Winograd kernels 
- Add new fused Winograd kernel 
- Enable weight flipping in Winograd kernels 

## Checklist

Put an `x` in the boxes that apply.

- [x] I have read the [CONTRIBUTING](https://github.com/ml-explore/mlx/blob/main/CONTRIBUTING.md) document
- [x] I have run `pre-commit run --all-files` to format my code / installed pre-commit prior to committing changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have updated the necessary documentation (if needed)

### CODE REVIEW

**Comment by jagrit06** [comment]
This update on its own should not help training benchmarks like cifar since the added kernel does not do well with large batch sizes. Further updates focused on batches should help that. In the meantime, this should improve batch size = 1 workloads and also reduce at least 1 copy of the inputs that might be used for padding 

The numbers below compare perf to PyTorch in the last column 
```
M3 Max Before:
(4, 128, 128, 32), (32, 3, 3, 32), float32, (1, 1), (1, 1), 0.343, -21.22%
(4, 128, 128, 64), (64, 3, 3, 64), float32, (1, 1), (1, 1), 1.013, -26.64%
(256, 16, 16, 32), (32, 3, 3, 32), float32, (1, 1), (1, 1), 0.352, -27.20%
(256, 8, 8, 64), (64, 3, 3, 64), float32, (1, 1), (1, 1), 0.340, -26.26%
(4, 128, 128, 128), (128, 3, 3, 128), float32, (1, 1), (1, 1), 2.788, -5.70%
(1, 16, 16, 32), (32, 3, 3, 32), float32, (1, 1), (1, 1), 0.074, +47.40%
(1, 16, 16, 64), (64, 3, 3, 64), float32, (1, 1), (1, 1), 0.106, -2.51%
(1, 128, 128, 128), (128, 3, 3, 128), float32, (1, 1), (1, 1), 0.780, -7.04%
(1, 128, 128, 256), (256, 3, 3, 256), float32, (1, 1), (1, 1), 2.261, +16.55%
(1, 16, 16, 128), (128, 3, 3, 128), float32, (1, 1), (1, 1), 0.157, -26.23%
(1, 16, 16, 256), (256, 3, 3, 256), float32, (1, 1), (1, 1), 0.274, -27.10%

M3 Max After:
(4, 128, 128, 32), (32, 3, 3, 32), float32, (1, 1), (1, 1), 0.342, -27.23%
(4, 128, 128, 64), (64, 3, 3, 64), float32, (1, 1), (1, 1), 1.014, -26.30%
(256, 16, 16, 32), (32, 3, 3, 32), float32, (1, 1), (1, 1), 0.346, -26.63%
(256, 8, 8, 64), (64, 3, 3, 64), float32, (1, 1), (1, 1), 0.340, -25.95%
(4, 128, 128, 128), (128, 3, 3, 128), float32, (1, 1), (1, 1), 2.179, +20.46%
(1, 16, 16, 32), (32, 3, 3, 32), float32, (1, 1), (1, 1), 0.052, +108.45%
(1, 16, 16, 64), (64, 3, 3, 64), float32, (1, 1), (1, 1), 0.059, +64.23%
(1, 128, 128, 128), (128, 3, 3, 128), float32, (1, 1), (1, 1), 0.692, +4.99%
(1, 128, 128, 256), (256, 3, 3, 256), float32, (1, 1), (1, 1), 1.634, +58.94%
(1, 16, 16, 128), (128, 3, 3, 128), float32, (1, 1), (1, 1), 0.082, +51.14%
(1, 16, 16, 256), (256, 3, 3, 256), float32, (1, 1), (1, 1), 0.158, +23.45%

M2 Ultra Before:
(4, 128, 128, 32), (32, 3, 3, 32), float32, (1, 1), (1, 1), 0.271, -27.74%
(4, 128, 128, 64), (64, 3, 3, 64), float32, (1, 1), (1, 1), 0.750, -35.50%
(256, 16, 16, 32), (32, 3, 3, 32), float32, (1, 1), (1, 1), 0.272, -25.79%
(256, 8, 8, 64), (64, 3, 3, 64), float32, (1, 1), (1, 1), 0.241, -22.92%
(4, 128, 128, 128), (128, 3, 3, 128), float32, (1, 1), (1, 1), 1.972, -23.75%
(1, 16, 16, 32), (32, 3, 3, 32), float32, (1, 1), (1, 1), 0.072, +45.15%
(1, 16, 16, 64), (64, 3, 3, 64), float32, (1, 1), (1, 1), 0.121, -15.60%
(1, 128, 128, 128), (128, 3, 3, 128), float32, (1, 1), (1, 1), 0.720, -32.91%
(1, 128, 128, 256), (256, 3, 3, 256), float32, (1, 1), (1, 1), 2.147, -24.63%
(1, 16, 16, 128), (128, 3, 3, 128), float32, (1, 1), (1, 1), 0.197, -32.59%
(1, 16, 16, 256), (256, 3, 3, 256), float32, (1, 1), (1, 1), 0.355, -39.01%

M2 Ultra After:
(4, 128, 128, 32), (32, 3, 3, 32), float32, (1, 1), (1, 1), 0.271, -27.77%
(4, 128, 128, 64), (64, 3, 3, 64), float32, (1, 1), (1, 1), 0.750, -35.65%
(256, 16, 16, 32), (32, 3, 3, 32), float32, (1, 1), (1, 1), 0.271, -25.83%
(256, 8, 8, 64), (64, 3, 3, 64), float32, (1, 1), (1, 1), 0.241, -23.95%
(4, 128, 128, 128), (128, 3, 3, 128), float32, (1, 1), (1, 1), 1.650, -8.85%
(1, 16, 16, 32), (32, 3, 3, 32), float32, (1, 1), (1, 1), 0.046, +122.89%
(1, 16, 16, 64), (64, 3, 3, 64), float32, (1, 1), (1, 1), 0.055, +84.06%
(1, 128, 128, 128), (128, 3, 3, 128), float32, (1, 1), (1, 1), 0.572, -15.53%
(1, 128, 128, 256), (256, 3, 3, 256), float32, (1, 1), (1, 1), 1.258, +28.53%
(1, 16, 16, 128), (128, 3, 3, 128), float32, (1, 1), (1, 1), 0.076, +67.96%
(1, 16, 16, 256), (256, 3, 3, 256), float32, (1, 1), (1, 1), 0.139, +55.04%
```

**Comment by awni** [comment]
I ran this [ResNet inference benchmark](https://gist.github.com/awni/dbb3e2a4bc66e1ed7fda3e9076d9886a) on M2 Ultra. 

Pre/post below. 

TLDR nice speedup on batch size = 1, slight slow down on 32+, is that expected?


Batch Size | Images-per-second | Milliseconds-per-image
---- | ---- | ---- 
1 | 608.598 | 1.643
2 | 1147.084 | 0.872
4 | 1426.277 | 0.701
8 | 1869.651 | 0.535
16 | 2133.113 | 0.469
32 | 2308.229 | 0.433
64 | 2529.965 | 0.395


Batch Size | Images-per-second | Milliseconds-per-image
---- | ---- | ---- 
1 | 747.178 | 1.338
2 | 1146.119 | 0.873
4 | 1539.774 | 0.649
8 | 1921.938 | 0.520
16 | 2035.330 | 0.491
32 | 2062.188 | 0.485
64 | 2106.746 | 0.475

**Comment by awni** [comment]
>  since the added kernel does not do well with large batch sizes. Further updates focused on batches should help that.

I guess that's what you meant. Should we just dispatch to the old kernel with batch size 32+?

**Comment by jagrit06** [comment]
> > since the added kernel does not do well with large batch sizes. Further updates focused on batches should help that.
> 
> I guess that's what you meant. Should we just dispatch to the old kernel with batch size 32+?

Done! I moved them back to the old routing for now

File: mlx/ops.cpp [hunk 1]
```diff
@@ -3864,6 +3864,21 @@ array conv_general(
       kernel_dilation,
       input_dilation);
 
+  // Check for direct conv
+  if (spatial_dims == 2 && groups == 1 && wt.shape(1) == 1 &&
+      wt.shape(2) == 1 && stride[0] == 1 && stride[1] == 1 &&
+      padding_lo[0] == 1 && padding_lo[1] == 1 && input_dilation[0] == 1 &&
+      input_dilation[1] == 1 && kernel_dilation[0] == 1 &&
+      kernel_dilation[1] == 1) {
```
Comment (angeloskath) [inline mlx/ops.cpp:3872]: This is nice!

I think there is an issue with the check. The padding should be 0 instead of 1. And what about flip? Also the following should generalize to all convs, no need to limit it to 2d.

```cpp
auto is_one = [](int x) { return x == 1; };
auto is_zero = [](int x) { return x == 0; };
if (groups == 1 &&
    wt.shape(0) * wt.shape(-1) == wt.size() && 
    std::all_of(stride.begin(), stride.end(), is_one) &&
    std::all_of(input_dilation.begin(), input_dilation.end(), is_one) &&
    std::all_of(kernel_dilation.begin(), kernel_dilation.end(), is_one) &&
    std::all_of(padding_lo.begin(), padding_lo.end(), is_zero) &&
    std::all_of(padding_hi.begin(), padding_hi.end(), is_zero)) {
  return matmul(in, transpose(flatten(wt, 0, -2, s), {1, 0}, s), s);
}
```
We could route in the `Convolution` primitive as well... maybe it makes a bit more sense 🤷‍♂️

File: mlx/backend/metal/kernels/conv.metal [hunk 2]
```diff
@@ -652,3 +687,373 @@ winograd_conv_2d_output_transform(
...
+  const bool is_edge_w_lo = iH_idx < 0;
+  const bool is_edge_h_lo = iW_idx < 0;
+  const bool is_edge_w_hi = iH_idx >= params.iS[0];
+  const bool is_edge_h_hi = iW_idx >= params.iS[1];
+  const bool is_edge =
+      is_edge_w_lo || is_edge_h_lo || is_edge_w_hi || is_edge_h_hi;
+
+  // Iterate over C
+  for (int c = 0; c < params.C; c += BC) {
+#define tmp_load_wt_idx(o, h, w, c) h* FA* BC* BO + w* BC* BO + c* BO + o
```
Comment (angeloskath) [inline mlx/backend/metal/kernels/conv.metal:808]: Why not implement these outside the kernel?

Also nitpick: `h* FA* BC` -> `h * FA * BC`

File: mlx/backend/metal/kernels/conv.metal [hunk 3]
```diff
@@ -445,14 +460,21 @@ winograd_conv_2d_input_transform(
...
+  const bool is_edge_h_lo = bh < 0;
+  const bool is_edge_w_hi = bw + (TW - 1) >= params.iS[0];
+  const bool is_edge_h_hi = bh + (TH - 1) >= params.iS[1];
+  const bool is_edge =
+      is_edge_w_lo || is_edge_h_lo || is_edge_w_hi || is_edge_h_hi;
 
   // Move to the correct input tile
-  inp_in += tid.z * params.in_strides[0] + bh * params.in_strides[1] +
-      bw * params.in_strides[2];
+  inp_in += tid.z * params.in_strides[0] + bh * int64_t(params.in_strides[1]) +
```
Comment (angeloskath) [inline mlx/backend/metal/kernels/conv.metal:476]: Why do you need the casting here I wonder. `param.in_strides` is already `int64_t` isn't it?

File: mlx/backend/metal/kernels/conv.metal [hunk 4]
```diff
@@ -652,3 +687,373 @@ winograd_conv_2d_output_transform(
 instantiate_winograd_conv_2d(float32, float);
 instantiate_winograd_conv_2d(bfloat16, bfloat16_t);
 instantiate_winograd_conv_2d(float16, half); // clang-format on
+
+#include "mlx/backend/metal/kernels/steel/attn/mma.h"
```
Comment (angeloskath) [inline mlx/backend/metal/kernels/conv.metal:691]: Which code is using `mlx/backend/metal/kernels/steel/gemm/mma.h` and should we move it to this?

File: mlx/backend/metal/kernels/conv.metal [hunk 5]
```diff
@@ -652,3 +687,373 @@ winograd_conv_2d_output_transform(
...
+
+  const int oH_idx = FN * (t_idx / tWu);
+  const int oW_idx = FN * (t_idx % tWu);
+  const int iH_idx = oH_idx + thr_h - params.pad[0];
+  const int iW_idx = oW_idx + thr_w - params.pad[1];
+
+  // Move to correct location
+
+  // clang-format off
+  input +=  b_idx * int64_t(params.in_strides[0]) + // N
```
Comment (angeloskath) [inline mlx/backend/metal/kernels/conv.metal:789]: Same as above.

**Review by angeloskath — COMMENTED** [review]
Awesome job as always :-) The MMATiles result in very readable kernels.

I left a few comments with the main one being an issue in the op with routing to matmul and the rest are either nitpicks or discussion.

File: mlx/backend/metal/kernels/conv.metal (hunk 4 above)
Comment (jagrit06) [reply mlx/backend/metal/kernels/conv.metal:691]: Yes, we should at some point move all of them under the same umbrella 
I plan to update the gems in the future

File: mlx/backend/metal/conv.cpp [hunk 6]
```diff
@@ -876,6 +858,39 @@ void Convolution::eval_gpu(const std::vector<array>& inputs, array& out) {
...
+        /*ldd = */ wt.shape(0),
+        /*transpose_a = */ false,
+        /*transpose_b = */ true,
+        /*batch_shape = */ {1},
+        /*batch_strides = */ {1},
+        /*A_batch_stride = */ 0,
+        /*B_batch_stride = */ 0,
+        /*matrix_stride_out = */ 0,
+        /*copies = */ empty_copies);
+  }
```
Comment (angeloskath) [inline mlx/backend/metal/conv.cpp:892]: I would say the `if` below needs an `else if`

**Review by angeloskath — APPROVED** [review]
Left a comment on the routing to matmul. Otherwise looks great!
